src/runner/browser-set.js

```diff
@@ -0,0 +1,155 @@
+                resolve();
+            });
+
+            bc.once('idle', eventHandler);
+            bc.once('closed', eventHandler);
+        });
+    }
+
+    _closeConnection (bc) {
+        return new Promise(resolve => {
+            if (bc.closed) {
+                resolve();
+                return;
+            }
+
+            bc.close();
+            bc.once('closed', resolve);
+        });
+    }
+
+    _createConnectionsReadyPromise () {
+        var ready = this.connections
+            .filter(bc => !bc.ready)
+            .map(bc => new Promise(resolve => bc.once('ready', resolve)));
+
+        return Promise.all(ready);
+    }
+
+    _waitConnectionsReady () {
+        return new Promise((resolve, reject) => {
+            var timeout = setTimeout(() => {
+                reject(new Error(getText(MESSAGE.cantEstablishBrowserConnection)));
+            }, this.BROWSER_CONNECTION_READY_TIMEOUT);
+
+            function errorHandler (msg) {
+                clearTimeout(timeout);
+                reject(new Error(msg));
+            }
+
+            this.once('error', errorHandler);
+
+            this
+                ._createConnectionsReadyPromise()
+                .then(() => {
+                    clearTimeout(timeout);
+                    this.removeListener('error', errorHandler);
+                    resolve();
+                });
+        });
+    }
+
+    async _checkForDisconnections () {
+        var disconnectedUserAgents = this.connections
+            .filter(bc => bc.closed)
+            .map(bc => bc.userAgent);
+
+        if (disconnectedUserAgents.length) {
+            await this.dispose();
+
+            throw new Error(getText(MESSAGE.cantRunAgainstDisconnectedBrowsers, disconnectedUserAgents.join(', ')));
+        }
+    }
+
+    _error (msg) {
+        this.emit('error', msg);
+
+        this.dispose();
+    }
+
+    //API
+    static async createBrowserSet (connections) {
+        var browserSet = new BrowserSet();
+
+        browserSet.connections = browserSet.connections.concat(connections);
+
+        var errorHandler = msg => browserSet._error(msg);
+
+        connections.forEach(bc => {
+            bc.on('error', errorHandler);
+
+            var freeErrorHandler = once(() => bc.removeListener('error', errorHandler));
+
+            bc.once('idle', freeErrorHandler);
+            bc.once('closed', freeErrorHandler);
+        });
+
+        await browserSet._checkForDisconnections();
+        await browserSet._waitConnectionsReady();
+
+        return browserSet;
+    }
+
+    freeConnection (bc) {
+        var index = this.connections.indexOf(bc);
+
+        if (index === -1)
+            return Promise.resolve();
+
+        var promisedAction = bc instanceof LocalBrowserConnection ? this._closeConnection(bc) : this._waitIdle(bc);
+
+        promisedAction = promisedAction
+            .then(() => {
+                remove(this.connections, bc);
```

Comment:
If we'll remove freed connection instantly before performing any action you will not need `promisedAction`s